src/models/PartialFormSubmission.php

```diff
@@ -93,123 +145,157 @@ public function getCMSFields()
+    /**
+     * Get the share link of the form
+     *
+     * @return string
+     * @throws Exception
+     */
+    public function getPartialLink()
     {
-        return $this->UserDefinedForm();
+        if (!$this->isInDB()) {
+            return '(none)';
+        }
+
+        $token = $this->Token;
+
+        return Controller::join_links(
+            Director::absoluteBaseURL(),
+            'partial',
+            $this->generateKey($token),
+            $token
+        );
     }
 
     /**
-     * @param Member $member
-     * @param array $context
-     * @return bool
+     * Generate a key based on the share token salt
+     *
+     * @param string $token
+     * @return mixed
      */
-    public function canCreate($member = null, $context = [])
+    public function generateKey($token)
     {
-        return false;
+        return hash_pbkdf2('sha256', $token, $this->TokenSalt, 1000, 16);
     }
 
     /**
-     * @param Member
-     *
-     * @return boolean|string
+     * Generate the partial tokens
+     * If the submission is password protected, generate a password.
+     * @throws Exception
      */
-    public function canView($member = null)
+    public function onBeforeWrite()
     {
-        if ($this->UserDefinedForm()) {
-            return $this->UserDefinedForm()->canView($member);
-        }
-
-        return parent::canView($member);
+        parent::onBeforeWrite();
+        $this->getPartialToken();
+        $this->generatePassword();
     }
 
     /**
-     * @param Member
+     * Get the unique token for the share link
      *
-     * @return boolean|string
+     * @return bool|string|null
+     * @throws Exception
      */
-    public function canEdit($member = null)
+    protected function getPartialToken()
     {
-        if ($this->UserDefinedForm()) {
-            return $this->UserDefinedForm()->canEdit($member);
+        if (!$this->TokenSalt) {
+            $this->TokenSalt = $this->generateToken();
+            $this->Token = $this->generateToken();
         }
 
-        return parent::canEdit($member);
+        return $this->Token;
     }
 
     /**
-     * @param Member
+     * Generate a new token
      *
-     * @return boolean|string
+     * @return bool|string
+     * @throws Exception
      */
-    public function canDelete($member = null)
+    protected function generateToken()
     {
-        if ($this->UserDefinedForm()) {
-            return $this->UserDefinedForm()->canDelete($member);
-        }
+        $generator = new RandomGenerator();
 
-        return parent::canDelete($member);
+        return substr($generator->randomToken('sha256'), 0, 16);
     }
 
     /**
-     * Get the share link of the form
-     *
-     * @return string
-     * @throws \Exception
+     * @return void
      */
-    public function getPartialLink()
+    protected function generatePassword()
     {
-        if (!$this->isInDB()) {
-            return '(none)';
+        if (!$this->Password &&
```

Comment:
Can we keep this method for generating a password only and move the conditions outside of this function? Also set the session in the controller, not here.

Reply:
The password needs to be set in the session there, otherwise it's entirely gone and not retrievable anymore.

The whole check for parent is not really necessary, just setting a password anyway doesn't hurt